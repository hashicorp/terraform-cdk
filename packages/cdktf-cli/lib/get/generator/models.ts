import { Schema } from './provider-schema';

export class Scope {
  constructor(public readonly name: string, public isComputed = false) {}
}

export interface AttributeModelOptions {
  storageName: string; // private property
  name: string;
  type: AttributeTypeModel;
  optional: boolean;
  computed: boolean;
  terraformName: string;
  terraformFullName: string;
  description?: string;
  getAttCall?: string;
}

export class AttributeModel {
  public storageName: string; // private property
  public name: string;
  public type: AttributeTypeModel;
  public optional: boolean;
  public computed: boolean;
  public terraformName: string;
  public terraformFullName: string;
  public description?: string;
  public getAttCall?: string;

  constructor(options: AttributeModelOptions) {
    this.storageName = options.storageName;
    this.name = options.name;
    this.type = options.type;
    this.optional = options.optional;
    this.computed = options.computed;
    this.terraformName = options.terraformName;
    this.terraformFullName = options.terraformFullName;
    this.description = options.description;
    this.getAttCall = options.getAttCall;
  }

  public get typeDefinition() {
    const optional = this.optional ? '?' : '';
    return `${this.name}${optional}: ${this.type.type}`;
  }

  public getterFunctionHeader() {
    return `public get ${this.name}()`
  }

  public getterFunctionBody() {
    if (this.getAttCall) return `return ${this.type.determineGetAttCall(this.name)};`;
    if (!this.getAttCall) return `return this.${this.storageName};`;
    return 'throw new Error("not implemented");'
  }

  public get isAssignable() {
    return !this.computed
  }

  public get isOptional() {
    return this.optional
  }
}

interface ResourceModelOptions {
  terraformType: string;
  className: string;
  baseName: string;
  fileName: string;
  filePath: string;
  attributes: AttributeModel[];
  structs: Struct[];
  provider: string;
  schema: Schema;
}

export class ResourceModel {
  public terraformType: string;
  public className: string;
  public baseName: string;
  public provider: string;
  public fileName: string;
  public filePath: string;
  public attributes: AttributeModel[];
  public schema: Schema;
  private _structs: Struct[];
  private dependencies: string[]

  constructor(options: ResourceModelOptions) {
    this.terraformType = options.terraformType
    this.className = options.className
    this.baseName = options.baseName
    this.attributes = options.attributes
    this.schema = options.schema
    this.provider = options.provider
    this.fileName = options.fileName;
    this.filePath = options.filePath;
    this._structs = options.structs

    this.dependencies = [
      `import { Construct } from 'constructs';`,
      `import { TerraformResource } from 'cdktf';`
    ]
  }

  public get structs(): Struct[] {
    return [this.configStruct, ...this._structs]
  }

  public get configStruct() {
    return new Struct(`${this.className}Config`, this.attributes)
  }

  public get assignableAttributes(): AttributeModel[] {
    return this.attributes.filter(attribute => !attribute.computed)
  }

  public get importStatements(): string[] {
    return this.dependencies
  }

  public get schemaAsJson(): string {
    return JSON.stringify(this.schema, undefined, 2)
  }
}

export class Struct {
  constructor(public readonly name: string, public readonly attributes: AttributeModel[], public readonly isComputed = false) {}

  public attributeName(attributeName: string): string {
    return `${this.allAttributesComputed ? '_' : ''}${attributeName}`;
  }

  public get assignableAttributes(): AttributeModel[] {
    return this.attributes.filter(attribute => !attribute.computed)
  }

  private get allAttributesOptional(): boolean {
    return this.requiredAttributes.length > 0
  }

  private get allAttributesComputed(): boolean {
    const computedAttributes = this.attributes.filter(a => a.computed)
    return computedAttributes.length === this.attributes.length
  }

  private get requiredAttributes(): AttributeModel[] {
    return this.attributes.filter(a => !a.optional && !a.computed);
  }

  public get attributeType() {
    return `${this.name}${this.allAttributesOptional ? ' = {}' : ''}`
  }
}

export class AttributeTypeModelOptions {
  public struct?: Struct;
  public isList?: boolean;
  public isComputed?: boolean;
  public isMap?: boolean;
}

export enum TokenizableTypes {
  STRING = 'string',
  NUMBER = 'number'
}

export interface ComputedComplexListOptions {
  name: string;
  type: string;
}

export class AttributeTypeModel {
  public isList: boolean;
  public isComputed: boolean;
  public isMap: boolean;
  public struct?: Struct;

  constructor(private _type: string, options: AttributeTypeModelOptions = {}) {
    this.isList = !!options.isList;
    this.isMap = !!options.isMap;
    this.isComputed = !!options.isComputed;
    this.struct = options.struct;
  }

  public get type(): string {
    if (this.isMap) return `{ [key: string]: ${this._type} }`;
    if (this.isList && !this.isComputed) return `${this._type}[]`;
    if (this.isList && this.isComputed && this.isPrimitive) return `${this._type}[]`;
    if (this.isList && this.isComputed && this.isComplex) return `${this._type}`;
    return this._type
  }

  public determineGetAttCall(terraformAttributeName: string): string {
    if (this.type === TokenizableTypes.STRING) {
      return this.isList ? this.getListAttribute(terraformAttributeName) : this.getStringAttribute(terraformAttributeName)
    }
    if (this.type === TokenizableTypes.NUMBER) { return this.getNumberAttribute(terraformAttributeName) }
    return 'any'
  }

  public computedComplexList(argument: ComputedComplexListOptions): string {
    return `new ${this.type}(${argument.name})`
  }

  public get isComplex(): boolean {
    return !!this.struct
  }

  public get isPrimitive(): boolean {
    return !this.isComplex
  }

  public get isComputedComplexList(): boolean {
    return this.isList && this.isComputed && this.isComplex
  }

  public get isInterpolatable(): boolean {
    return Object.values(TokenizableTypes).includes(this.type as TokenizableTypes)
  }

  private getListAttribute(name: string): string {
    return `this.getListAttribute('${name}')`
  }

  private getStringAttribute(name: string): string {
    return `this.getStringAttribute('${name}')`
  }

  private getNumberAttribute(name: string): string {
    return `this.getNumberAttribute('${name}')`
  }
}