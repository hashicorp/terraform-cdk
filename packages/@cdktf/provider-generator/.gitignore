**/*.d.ts.map
**/*.js.map
tsconfig.tsbuildinfo
!tsconfig.json
edge-provider-bindings
cdktf-provider-generator-*.tgz
!test/generate-provider-tests.js
!ambient.d.ts