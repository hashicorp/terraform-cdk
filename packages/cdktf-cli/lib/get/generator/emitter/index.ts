export * from './attributes-emitter'
export * from './resource-emitter'
export * from './struct-emitter'