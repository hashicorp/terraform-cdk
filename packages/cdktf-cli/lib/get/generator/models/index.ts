export * from "./attribute-model"
export * from "./attribute-type-model"
export * from "./resource-model"
export * from "./scope"
export * from "./struct"