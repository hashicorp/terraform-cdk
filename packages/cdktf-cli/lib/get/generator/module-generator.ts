import { CodeMaker, toCamelCase } from "codemaker";
import { Module, Submodule } from "../registry-client";

export class ModuleGenerator {
  constructor(private readonly code: CodeMaker, private readonly spec: Module) {
    this.code.indentation = 2;
    this.emitSubmodule(spec.root);
  }

  private emitSubmodule(spec: Submodule) {
    const source = `${this.spec.namespace}/${this.spec.name}/${this.spec.provider}`;
    const fileName = `modules/${source}.ts`;
    this.code.openFile(fileName);

    this.code.line(`// generated by cdktf get`);
    this.code.line(`// ${this.spec.id}/${spec.path}`);

    this.code.line(`import { TerraformModule } from 'cdktf';`);
    this.code.line(`import { Construct } from 'constructs';`);

    const baseName = this.code.toPascalCase(spec.name.replace(/-/g, '_'));
    const optionsType = `${baseName}Options`;

    this.code.openBlock(`export interface ${optionsType}`);
    for (const input of spec.inputs) {
      const optional = (input.required && (input.default === undefined)) ? '' : '?';
      this.code.line(`/**`);
      this.code.line(` * ${input.description}`);
      if (input.default) {
        this.code.line(` * @default ${input.default}`);
      }
      this.code.line(` */`);
      this.code.line(`readonly ${toCamelCase(input.name)}${optional}: ${parseType(input.type)};`);
    }
    this.code.closeBlock();

    this.code.openBlock(`export class ${baseName} extends TerraformModule`);

    this.code.line(`private readonly inputs: { [name: string]: any } = { }`);

    const allOptional = spec.inputs.find(x => x.required) ? '' : ' = {}';

    this.code.open(`public constructor(scope: Construct, id: string, options: ${optionsType}${allOptional}) {`);
    this.code.open(`super(scope, id, {`);
    this.code.line(`source: '${source}',`);
    this.code.line(`version: '${this.spec.version}',`);
    this.code.close(`});`);

    for (const input of spec.inputs) {
      const inputName = toCamelCase(input.name);
      this.code.line(`this.${inputName} = options.${inputName};`);
    }

    this.code.close(`}`); // ctor

    for (const input of spec.inputs) {
      const inputName = toCamelCase(input.name);
      const inputType = parseType(input.type) + ((input.required && (input.default === undefined)) ? '' : ' | undefined');
      this.code.openBlock(`public get ${inputName}(): ${inputType}`);
      this.code.line(`return this.inputs['${input.name}'] as ${inputType};`);
      this.code.closeBlock();

      this.code.openBlock(`public set ${inputName}(value: ${inputType})`);
      this.code.line(`this.inputs['${input.name}'] = value;`);
      this.code.closeBlock();
    }

    this.code.openBlock(`protected synthesizeAttributes()`);
    this.code.line(`return this.inputs;`);
    this.code.closeBlock();

    this.code.closeBlock(); // class

    this.code.closeFile(fileName);
  }

}

function parseType(type: string) {
  if (type === 'string') { return 'string'; }
  if (type === 'number') { return 'number'; }
  if (type === 'bool') { return 'boolean'; }
  if (type === 'list') { return 'string[]'; }
  if (type === 'map') { return '{ [key: string]: string }'; }
  if (type === 'any') { return 'any'; }

  const complexType = parseComplexType(type);
  if (complexType) { return complexType; }

  throw new Error(`unknown type ${type}`);
}

function parseComplexType(type: string): string | undefined {
  const complex = /^(object|list|map)\(([\s\S]+)\)/;
  const match = complex.exec(type);
  if (!match) {
    return undefined;
  }

  const [ , kind, innerType ] = match;

  if (kind === 'object') {
    return `any`;
  }

  if (kind === 'list') {
    return `${parseType(innerType)}[]`;
  }

  if (kind === 'map') {
    return `{ [key: string]: ${parseType(innerType) } }`;
  }

  throw new Error(`unexpected kind ${kind}`);
}