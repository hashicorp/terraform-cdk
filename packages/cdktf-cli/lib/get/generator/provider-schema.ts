import * as fs from 'fs-extra';
import * as path from 'path';
import { spawn, SpawnOptions } from 'child_process';
import { promisify } from 'util';
import { withTempDir } from '../../util';

const writeFile = promisify(fs.writeFile);
const mkdirp = promisify(fs.mkdirp);

export interface ProviderSchema {
  format_version: '1.0';
  provider_schemas?: { [type: string]: Provider };
}

export interface Provider {
  provider: Schema;
  resource_schemas: { [type: string]: Schema };
  data_source_schemas: { [type: string]: Schema };
}

export interface Schema {
  version: number;
  block: Block;
}

export interface Attribute {
  type: AttributeType;
  description?: string;
  required?: boolean;
  optional?: boolean;
  computed?: boolean;
  sensitive?: boolean;
}

export type AttributeType =
  'string'
  | 'bool'
  | 'number'
  | [ 'set', AttributeType ]
  | [ 'map', AttributeType ]
  | [ 'list', AttributeType ]
  | [ 'object', { [attribute: string]: AttributeType } ];

export type BlockType = {
  nesting_mode: 'single' | 'map';
  block: Block;
} | {
  nesting_mode: 'list' | 'set';
  block: Block;
  min_items?: number;
  max_items?: number;
}

export interface Block {
  attributes: { [name: string]: Attribute };
  block_types: { [name: string]: BlockType };
}

export async function readSchema(providers: string[]): Promise<ProviderSchema> {
  const provider: { [name: string]: { version?: string } } = { };
  for (const p of providers) {
    const [ name, version ] = p.split('@');
    provider[name] = { version };
  }
  let schema = '';
  const workDir = process.cwd()

  await withTempDir('fetchSchema', async () => {
    const outdir = process.cwd();
    const filePath = path.join(outdir, 'providers.tf.json');
    await writeFile(filePath, JSON.stringify({ provider }));

    const env = process.env['TF_PLUGIN_CACHE_DIR'] ? process.env : Object.assign({}, process.env, { 'TF_PLUGIN_CACHE_DIR': await cacheDir(workDir) })

    // todo: when implementing logging, we need to make sure we can show the terraform init
    // output if the log level is set to debug
    await exec('terraform', [ 'init' ], { cwd: outdir, env });
    schema = await exec('terraform', ['providers', 'schema', '-json'], { cwd: outdir, env });
    fs.unlinkSync(filePath)
  })

  return JSON.parse(schema);
}

async function cacheDir(workDir: string) {
  const cacheDir = path.join(workDir, '.terraform/plugins');
  await mkdirp(cacheDir);
  return cacheDir
}

// todo: move this exec function into terraform class and it should be shared by all
// terraform comamnds.
async function exec(command: string, args: string[], options: SpawnOptions = {}): Promise<string> {
  return new Promise((ok, ko) => {
    const child = spawn(command, args, options);
    const out = new Array<Buffer>();
    child.stdout?.on('data', (chunk: Buffer) => out.push(chunk));
    child.stderr?.on('data', (chunk: string | Uint8Array) => process.stderr.write(chunk));
    child.once('error', (err: any) => ko(err));
    child.once('close', (code: number) => {
      if (code !== 0) {
        return ko(new Error(`non-zero exit code ${code}`));
      }

      return ok(Buffer.concat(out).toString('utf-8'));
    });
  });
}
