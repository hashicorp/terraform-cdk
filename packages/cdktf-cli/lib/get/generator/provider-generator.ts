import { CodeMaker, toCamelCase, toPascalCase, toSnakeCase } from 'codemaker';
import { Attribute, AttributeType, Block, BlockType, Provider, ProviderSchema, Schema } from './provider-schema';

export class TerraformGenerator {
  private anonymousStructs = new Array<Struct>();
  private classNames: string[] = [];

  constructor(private readonly code: CodeMaker, schema: ProviderSchema) {
    this.code.indentation = 2;

    if (!schema.provider_schemas) {
      console.error('warning: no providers');
      return;
    }

    for (const [name, provider] of Object.entries(schema.provider_schemas)) {
      this.emitProvider(name, provider);
    }
  }

  public async save(outdir: string) {
    await this.code.save(outdir);
  }

  private emitProvider(name: string, provider: Provider) {
    const files: string[] = []
    for (const [type, resource] of Object.entries(provider.resource_schemas)) {
      files.push(this.emitResource(name, type, resource));
    }
    this.emitIndexFile(name, files)
  }

  private emitIndexFile(provider: string, files: string[]): void {
    const folder = `providers/${provider}`
    const filePath = `${folder}/index.ts`
    this.code.openFile(filePath);
    this.code.line('// generated by cdktf get')
    for (const file of files) {
      this.code.line(`export * from './${file.replace(`${folder}/`, '').replace('.ts', '')}';`);
    }
    this.code.line()
    this.code.closeFile(filePath)
  }

  private emitResource(provider: string, type: string, schema: Schema): string {
    const resource = this.parseResource(provider, type, schema);

    const filePath = `providers/${provider}/${resource.fileName}`;

    this.code.openFile(filePath);

    this.code.line(`// generated from terraform resource schema`);
    this.code.line();
    this.code.line('/*');
    this.code.line(JSON.stringify(schema, undefined, 2));
    this.code.line('*/');
    this.code.line(`import { Construct } from 'constructs';`);
    this.code.line(`import { TerraformResource } from 'cdktf';`);

    this.code.line();
    this.code.line(`// Configuration`)
    this.code.line();
    this.emitStruct({ name: resource.configName, attributes: resource.attributes });

    for (const struct of this.anonymousStructs) {
      this.emitStruct(struct);
    }

    this.anonymousStructs = [];

    this.code.line();
    this.code.line(`// Resource`)
    this.code.line();
    this.code.openBlock(`export class ${resource.className} extends TerraformResource`);

    this.emitHeader('INITIALIZER');
    this.emitInitializer(resource);

    this.emitHeader('ATTRIBUTES');
    this.emitResourceAttributes(resource);

    // synthesis
    this.emitHeader('SYNTHESIS');
    this.emitResourceSynthesis(resource);

    this.code.closeBlock(); // construct

    this.code.closeFile(filePath);

    return filePath;
  }

  private emitHeader(title: string) {
    this.code.line();
    this.code.line('// ' + '='.repeat(title.length));
    this.code.line(`// ${title}`);
    this.code.line('// ' + '='.repeat(title.length));
  }

  private emitResourceSynthesis(resource: ResourceModel) {
    this.code.line();
    this.code.openBlock(`public synthesizeAttributes()`);
    this.code.open(`return {`);

    for (const att of resource.attributes) {
      if (att.computed) { continue; }
      this.code.line(`${att.terraformName}: this.${att.storageName},`);
    }

    this.code.close(`};`);
    this.code.closeBlock();
  }

  private emitResourceAttributes(resource: ResourceModel) {
    for (const att of resource.attributes) {
      this.code.line();
      this.code.line(`// ${att.terraformName}`);

      if (!att.computed) {
        this.emitAttribute(att);
      } else {
        this.emitComputedAttribute(att);
      }
    }
  }

  private emitAttribute(att: AttributeModel) {
    this.code.line(`private ${att.storageName}?: ${att.type};`);

    // if we dont have a getAtt call, we will emit an optional attribute, since "undefined"
    // indicates this value is not specified.
    if (!att.getAttCall) {
      this.code.openBlock(`public get ${att.name}()`);
      this.code.line(`return this.${att.storageName};`);
      this.code.closeBlock();

      this.code.openBlock(`public set ${att.name}(value: ${att.type} | undefined)`);
      this.code.line(`this.${att.storageName} = value;`);
      this.code.closeBlock();
      return;
    }

    // otherwise, there is always a value - it will either be the value explicitly set
    // or the late-bound value through interpolation.

    this.code.openBlock(`public get ${att.name}()`);
    this.code.line(`return this.${att.storageName} ?? ${att.getAttCall};`);
    this.code.closeBlock();

    this.code.openBlock(`public set ${att.name}(value: ${att.type})`);
    this.code.line(`this.${att.storageName} = value;`);
    this.code.closeBlock();
  }

  private emitComputedAttribute(att: AttributeModel) {
    this.code.openBlock(`public get ${att.name}()`);
    if (att.getAttCall) {
      this.code.line(`return ${att.getAttCall};`);
    } else {
      const message = `computed attribute "$${att.terraformFullName}" has unsupported type "${att.type}"`;
      this.code.line(`throw new Error('${message}');`);
      // console.error(`warning: ${message}`);
    }
    this.code.closeBlock();
  }

  private emitInitializer(resource: ResourceModel) {
    this.code.line();

    const requiredAttributes = resource.attributes.filter(a => !a.optional && !a.computed);
    const allOptionals = requiredAttributes.length > 0 ? '' : ` = {}`;
    // `computed` attributes are skipped in `emitStruct`. This causes empty interfaces (e.g. in aws provider `FmsAdminAccount`)
    const computedAttributes = resource.attributes.filter(a => a.computed)
    const emptyConfigInterface = (computedAttributes.length !== resource.attributes.length) ? '' : '_'

    this.code.openBlock(`public constructor(scope: Construct, id: string, ${emptyConfigInterface}config: ${resource.configName}${allOptionals})`);

    // invoke super ctor with the terraform resource type
    this.code.open(`super(scope, id, {`);
    this.code.line(`type: '${resource.terraformType}',`);
    this.code.close(`});`);

    // initialize config properties
    for (const att of resource.attributes) {
      if (att.computed) {
        continue;
      }

      this.code.line(`this.${att.storageName} = config.${att.name};`);
    }

    this.code.closeBlock();
  }

  private emitStruct(struct: Struct) {
    this.code.openBlock(`export interface ${struct.name}`);
    for (const att of struct.attributes) {
      // skip computed attributes
      if (att.computed) {
        continue;
      }

      if (att.description) {
        this.code.line(`/** ${att.description} */`);
      }

      this.code.line(`readonly ${this.renderAttributeProperty(att)};`);
    }
    this.code.closeBlock();
  }

  //
  // name conversions

  private renderAttributeProperty(attribute: AttributeModel) {
    const optional = attribute.optional ? '?' : '';
    return `${attribute.name}${optional}: ${attribute.type}`;
  }

  private renderAttributeType(scope: string[], attributeType: AttributeType): string {
    if (typeof(attributeType) === 'string') {
      switch (attributeType) {
        case 'bool': return 'boolean';
        case 'string': return 'string';
        case 'number': return 'number';
        default: throw new Error(`invalid primitive type ${attributeType}`);
      }
    }

    if (Array.isArray(attributeType)) {
      if (attributeType.length !== 2) {
        throw new Error(`unexpected array`);
      }

      const [ kind, type ] = attributeType;

      if (kind === 'set' || kind === 'list') {
        return this.renderAttributeType(scope, type as AttributeType) + '[]';
      }

      if (kind === 'map') {
        const valueType: string = this.renderAttributeType(scope, type as AttributeType);
        return `{ [key: string]: ${valueType} }`;
      }

      if (kind === 'object') {
        const objAttributes = type as { [name: string]: AttributeType };
        const attributes: { [name: string]: Attribute } = { };
        for (const [ name, type ] of Object.entries(objAttributes)) {
          attributes[name] = { type }
        }
        const struct = this.addAnonymousStruct(scope, attributes);
        return struct.name;
      }

      throw new Error(`unexpected kind ${kind} with ${JSON.stringify(type)}`);
    }

    throw new Error(`unknown type ${attributeType}`);
  }

  private renderAttributesForBlock(parentType: string, block: Block) {
    const attributes = new Array<AttributeModel>();

    for (const [ terraformAttributeName, att ] of Object.entries(block.attributes || { })) {
      const type = this.renderAttributeType([ parentType, terraformAttributeName ], att.type);
      const name = toCamelCase(terraformAttributeName);
      attributes.push({
        getAttCall: determineGetAttCall(type, terraformAttributeName),
        terraformFullName: `${parentType}.${terraformAttributeName}`,
        description: att.description,
        name,
        storageName: `_${name}`,
        computed: !!att.computed,
        optional: !att.computed && !!att.optional,
        terraformName: terraformAttributeName,
        type
      })
    }

    for (const [ blockTypeName, blockType ] of Object.entries(block.block_types || { })) {
      // create a struct for this block
      const blockStruct = this.addStruct([ parentType, blockTypeName ], this.renderAttributesForBlock(`${parentType}_${blockTypeName}`, blockType.block))

      // define the attribute
      attributes.push(attributeForBlockType(blockTypeName, blockType, blockStruct));
    }

    return attributes;

    function determineGetAttCall(type: string, terraformAttributeName: string) {
      if (type === 'string') { return `this.getStringAttribute('${terraformAttributeName}')`; }
      if (type === 'number') { return `this.getNumberAttribute('${terraformAttributeName}')`; }
      if (type === 'string[]') { return `this.getListAttribute('${terraformAttributeName}')`; }
      return undefined;
    }

    function attributeForBlockType(terraformName: string, blockType: BlockType, struct: Struct): AttributeModel {
      const name = toCamelCase(terraformName);

      switch (blockType.nesting_mode) {
        case 'single':
          return {
            name,
            terraformName,
            terraformFullName: terraformName,
            type: struct.name,
            description: `${terraformName} block`,
            storageName: `_${name}`,
            optional: !struct.attributes.some(x => !x.optional),
            computed: false,
          };

        case 'map':
          return {
            name,
            terraformName,
            terraformFullName: terraformName,
            type: `{ [key: string]: ${struct.name} }`,
            description: `${terraformName} block`,
            storageName: `_${name}`,
            optional: false,
            computed: false,
          };

        case 'list':
        case 'set':
          return {
            name,
            terraformName: terraformName,
            terraformFullName: terraformName,
            type: `${struct.name}[]`,
            description: `${terraformName} block`,
            storageName: `_${name}`,
            optional: blockType.min_items === undefined ? true : blockType.min_items < 1,
            computed: false,
          };
      }
    }
  }

  private parseResource(provider: string, terraformType: string, schema: Schema): ResourceModel {

    let baseName = terraformType;
    if (baseName.startsWith(`${provider}_`)) {
      baseName = baseName.substr(provider.length + 1);
    }

    const className = this.uniqueClassName(toPascalCase(baseName));
    const fileName = `${toSnakeCase(baseName).replace(/_/g, '-')}.ts`;
    const configName = `${className}Config`;

    const resource: ResourceModel = {
      terraformType,
      baseName,
      className,
      fileName,
      configName,
      attributes: this.renderAttributesForBlock(baseName, schema.block)
    }

    return resource;
  }

  private uniqueClassName(className: string): string {
    if (this.classNames.indexOf(className) >= 0) {
      className = `${className}A`
    }
    this.classNames.push(className)
    return className
  }

  private addAnonymousStruct(scope: string[], attrs: { [name: string]: Attribute }) {
    const attributes = new Array<AttributeModel>();
    for (const [ terraformName, att ] of Object.entries(attrs)) {
      const name = toCamelCase(terraformName);
      attributes.push({
        name,
        storageName: `_${name}`,
        computed: false,
        description: att.description,
        optional: true,
        terraformName,
        terraformFullName: [ ...scope, terraformName ].join('_'),
        type: this.renderAttributeType([ ...scope, terraformName ], att.type),
      });
    }

    return this.addStruct(scope, attributes);
  }

  private addStruct(scope: string[], attributes: AttributeModel[]) {
    const s = {
      name: this.uniqueClassName(toPascalCase(scope.map(x => toSnakeCase(x)).join('_'))),
      attributes
    }
    this.anonymousStructs.push(s);
    return s;
  }

}


interface Struct {
  readonly name: string;
  readonly attributes: AttributeModel[];
}

interface AttributeModel {
  storageName: string; // private property
  name: string;
  type: string;
  optional: boolean;
  computed: boolean;
  terraformName: string;
  terraformFullName: string;
  description?: string;
  getAttCall?: string;
}

interface ResourceModel {
  terraformType: string;
  className: string;
  baseName: string;
  fileName: string;
  configName: string;
  attributes: AttributeModel[];
}
