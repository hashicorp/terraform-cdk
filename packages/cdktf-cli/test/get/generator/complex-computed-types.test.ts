import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { TerraformGenerator } from '../../../lib/get/generator/provider-generator';
import { CodeMaker } from 'codemaker';

test('generate an acm certifacte resource with complex computed types', async () => {
  const code = new CodeMaker()
  const workdir = fs.mkdtempSync(path.join(os.tmpdir(), 'complex-computed-types.test'));
  const spec = JSON.parse(fs.readFileSync(path.join(__dirname, 'fixtures', 'aws_acm_certificate.test.fixture.json'), 'utf-8'));
  new TerraformGenerator(code, spec);
  await code.save(workdir);

  const output = fs.readFileSync(path.join(workdir, 'providers/aws/acm-certificate.ts'), 'utf-8');
  expect(output).toMatchSnapshot();
});