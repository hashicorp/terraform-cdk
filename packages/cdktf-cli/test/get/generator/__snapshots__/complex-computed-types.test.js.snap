// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate an acm certifacte resource with complex computed types 1`] = `
"// generated from terraform resource schema

/*
{
  \\"version\\": 0,
  \\"block\\": {
    \\"attributes\\": {
      \\"arn\\": {
        \\"type\\": \\"string\\",
        \\"computed\\": true
      },
      \\"certificate_authority_arn\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true
      },
      \\"certificate_body\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true
      },
      \\"certificate_chain\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true
      },
      \\"domain_name\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true,
        \\"computed\\": true
      },
      \\"domain_validation_options\\": {
        \\"type\\": [
          \\"list\\",
          [
            \\"object\\",
            {
              \\"domain_name\\": \\"string\\",
              \\"resource_record_name\\": \\"string\\",
              \\"resource_record_type\\": \\"string\\",
              \\"resource_record_value\\": \\"string\\"
            }
          ]
        ],
        \\"computed\\": true
      },
      \\"id\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true,
        \\"computed\\": true
      },
      \\"private_key\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true,
        \\"sensitive\\": true
      },
      \\"subject_alternative_names\\": {
        \\"type\\": [
          \\"list\\",
          \\"string\\"
        ],
        \\"optional\\": true,
        \\"computed\\": true
      },
      \\"tags\\": {
        \\"type\\": [
          \\"map\\",
          \\"string\\"
        ],
        \\"optional\\": true
      },
      \\"validation_emails\\": {
        \\"type\\": [
          \\"list\\",
          \\"string\\"
        ],
        \\"computed\\": true
      },
      \\"validation_method\\": {
        \\"type\\": \\"string\\",
        \\"optional\\": true,
        \\"computed\\": true
      }
    },
    \\"block_types\\": {
      \\"options\\": {
        \\"nesting_mode\\": \\"list\\",
        \\"block\\": {
          \\"attributes\\": {
            \\"certificate_transparency_logging_preference\\": {
              \\"type\\": \\"string\\",
              \\"optional\\": true
            }
          }
        },
        \\"max_items\\": 1
      }
    }
  }
}
*/
import { Construct } from 'constructs';
import { TerraformResource } from 'cdktf';

// Configuration

export interface AcmCertificateConfig {
  readonly certificateAuthorityArn?: string;
  readonly certificateBody?: string;
  readonly certificateChain?: string;
  readonly privateKey?: string;
  readonly tags?: { [key: string]: string };
  /** options block */
  readonly options?: AcmCertificateOptions[];
}
export interface AcmCertificateDomainValidationOptions {
  readonly domainName?: string;
  readonly resourceRecordName?: string;
  readonly resourceRecordType?: string;
  readonly resourceRecordValue?: string;
}
export interface AcmCertificateOptions {
  readonly certificateTransparencyLoggingPreference?: string;
}

// Resource

export class AcmCertificate extends TerraformResource {

  // ===========
  // INITIALIZER
  // ===========

  public constructor(scope: Construct, id: string, config: AcmCertificateConfig = {}) {
    super(scope, id, {
      type: 'aws_acm_certificate',
    });
    this._certificateAuthorityArn = config.certificateAuthorityArn;
    this._certificateBody = config.certificateBody;
    this._certificateChain = config.certificateChain;
    this._privateKey = config.privateKey;
    this._tags = config.tags;
    this._options = config.options;
  }

  // ==========
  // ATTRIBUTES
  // ==========

  // arn
  public get arn() {
    return this.getStringAttribute('arn');
  }

  // certificate_authority_arn
  private _certificateAuthorityArn?: string;
  public get certificateAuthorityArn() {
    return this._certificateAuthorityArn ?? this.getStringAttribute('certificate_authority_arn');
  }
  public set certificateAuthorityArn(value: string) {
    this._certificateAuthorityArn = value;
  }

  // certificate_body
  private _certificateBody?: string;
  public get certificateBody() {
    return this._certificateBody ?? this.getStringAttribute('certificate_body');
  }
  public set certificateBody(value: string) {
    this._certificateBody = value;
  }

  // certificate_chain
  private _certificateChain?: string;
  public get certificateChain() {
    return this._certificateChain ?? this.getStringAttribute('certificate_chain');
  }
  public set certificateChain(value: string) {
    this._certificateChain = value;
  }

  // domain_name
  public get domainName() {
    return this.getStringAttribute('domain_name');
  }

  // domain_validation_options
  public get domainValidationOptions() {
    throw new Error('computed attribute \\"$acm_certificate.domain_validation_options\\" has unsupported type \\"AcmCertificateDomainValidationOptions[]\\"');
  }

  // id
  public get id() {
    return this.getStringAttribute('id');
  }

  // private_key
  private _privateKey?: string;
  public get privateKey() {
    return this._privateKey ?? this.getStringAttribute('private_key');
  }
  public set privateKey(value: string) {
    this._privateKey = value;
  }

  // subject_alternative_names
  public get subjectAlternativeNames() {
    return this.getListAttribute('subject_alternative_names');
  }

  // tags
  private _tags?: { [key: string]: string };
  public get tags() {
    return this._tags;
  }
  public set tags(value: { [key: string]: string } | undefined) {
    this._tags = value;
  }

  // validation_emails
  public get validationEmails() {
    return this.getListAttribute('validation_emails');
  }

  // validation_method
  public get validationMethod() {
    return this.getStringAttribute('validation_method');
  }

  // options
  private _options?: AcmCertificateOptions[];
  public get options() {
    return this._options;
  }
  public set options(value: AcmCertificateOptions[] | undefined) {
    this._options = value;
  }

  // =========
  // SYNTHESIS
  // =========

  protected synthesizeAttributes() {
    return {
      certificate_authority_arn: this._certificateAuthorityArn,
      certificate_body: this._certificateBody,
      certificate_chain: this._certificateChain,
      private_key: this._privateKey,
      tags: this._tags,
      options: this._options,
    };
  }
}
"
`;
